File: skills/fc-authoring/example-clinic-fc/protocols/plugins/plugin_bp.py
# Blood Pressure report -- a Tag.bio Python plugin.
# A plugin receives TagbioData (tag_data.df is a pandas DataFrame of the analysis frame:
# one row per entity, columns = the analysis_variables) and writes to TagbioResult.
from tagbiopy.protocol import TagbioData, TagbioResult
import plotly.express as px


def blood_pressure_report(tag_data: TagbioData, tag_result: TagbioResult):
    # A numeric variable arrives named "<Collection>: <Variable>" (e.g. "Blood Pressure:
    # Systolic" — the Python SDK uses ": "; the R SDK uses " = "). Drop the collection prefix so
    # the columns read as plain Systolic / Diastolic.
    df = tag_data.df.rename(columns=lambda c: c.replace("Blood Pressure: ", ""))
    tall = df.melt(
        id_vars=["Department"],
        value_vars=["Systolic", "Diastolic"],
        var_name="measure",
        value_name="mmHg",
    )
    fig = px.box(tall, x="Department", y="mmHg", color="measure")
    fig.write_html(tag_result.path)
    return tag_result

File: skills/fc-authoring/example-clinic-fc/protocols/plugins/plugin_bp.R
# Blood Pressure by Department -- a Tag.bio R plugin.
# A plugin is a function(tag_data, tag_result): pull the analysis frame with
# tagbio::get_results(), do the work, write to tag_result$output_path, return tag_result.
if (!require('plotly')) install.packages('plotly', repos = "https://cloud.r-project.org")
require('plotly')
require('dplyr')
require('tidyr')

function(tag_data, tag_result) {

  # One row per entity; columns are the analysis_variables (Encounter ID is used as the row name).
  data <- tagbio::get_results(tag_data, row_name = "Encounter ID")

  # A numeric variable arrives named "<Collection> = <Variable>" (e.g. "Blood Pressure = Systolic");
  # drop the collection prefix so the columns read as plain Systolic / Diastolic.
  names(data) <- sub("^Blood Pressure = ", "", names(data))

  data_tall <- data %>%
    tidyr::pivot_longer(c(Systolic, Diastolic), names_to = "measure", values_to = "mmHg")

  fig <- plot_ly(data_tall, x = ~Department, y = ~mmHg, color = ~measure, type = "box") %>%
    layout(boxmode = "group", yaxis = list(title = "mmHg"))

  htmlwidgets::saveWidget(fig, tag_result$output_path)

  return(tag_result)
}
